include/avif/avif_cxx.h

```diff
@@ -0,0 +1,35 @@
+// Copyright 2023 Google LLC
+// SPDX-License-Identifier: BSD-2-Clause
+
+// This is a header only library.
+
+#ifndef AVIF_AVIF_CXX_H
+#define AVIF_AVIF_CXX_H
+
+#if !(defined(__cplusplus) || defined(c_plusplus))
+#error "This a C++ only header. Use avif/avif.h for C."
+#endif
+
+#include <memory>
+
+#include "avif/avif.h"
+
+namespace libavif
+{
+
+// Struct to call the destroy functions in a unique_ptr.
+struct UniquePtrDeleter
+{
+    void operator()(avifEncoder * encoder) const { avifEncoderDestroy(encoder); }
+    void operator()(avifDecoder * decoder) const { avifDecoderDestroy(decoder); }
+    void operator()(avifImage * image) const { avifImageDestroy(image); }
+};
+
+// Use these unique_ptr to ensure the structs are automatically destroyed.
+typedef std::unique_ptr<avifEncoder, UniquePtrDeleter> EncoderPtr;
+typedef std::unique_ptr<avifDecoder, UniquePtrDeleter> DecoderPtr;
+typedef std::unique_ptr<avifImage, UniquePtrDeleter> ImagePtr;
```

Comment:
Optional: I am tempted to suggest we still add an "avif" or "Avif" prefix to these type names.  The reason is that we don't always use the fully qualified names `libavif::EncoderPtr`, `libavif::DecoderPtr`, etc. For example, we may write code inside a `namespace libavif` block:

```
namespace libavif {
  
  ...

  EncoderPtr encoder(avifEncoderCreate());
  ImagePtr image(avifImageCreateEmpty());
 
   ...

}  // namespace libavif
```

Reply:
Also, do we need the Ptr suffix for these variables?

Reply:
I like the `Ptr` suffix because it suggests these types are std::unique_ptr's. We can also use the `UniquePtr` suffix, but it's a little verbose.

Reply:
So Ptr is because we "might" (with big quotes because we have no idea whether it would be better, though we have done it for RGBImage and RWData in aviftest_helpers.h) create a C++ class for an encoder and we would name it `Encoder`. 
For the prefix, I think it is too verbose: either you develop within the libavif namespace (which is only done in tests) and therefore the libavif repo,  and you therefore know what you are doing and ImagePtr is obviously for an avifImage. Either you develop somewhere else, and `libavif::ImagePtr` is simply shorter and less redundant than `libavif::avifImagePtr` IMHO. WDYT?

Reply:
Yes, let's optimize for the fully-qualified `libavif::ImagePtr` case.

Note: I wonder if we can use the `avif` namespace. `avif::ImagePtr` looks really nice. libheif's heif_cxx.h header uses the `heif` namespace. Just a thought.

Reply:
I love the `avif` namespace: it is short and I doubt anybody uses it. Let me switch the tests to using that namespace: I am modifying all the files anyway.

Reply:
Done

Reply:
Good. As the r[eference implementation of AVIF](https://github.com/AOMediaCodec/av1-avif/wiki#reference-software), libavif should be qualified to use the `avif` namespace.